CadRevealFbxProvider/BatchUtils/ScaffoldOptimizer/ReplacementScaffoldParts/PrimitiveGeometryDetector.cs

```diff
@@ -0,0 +1,463 @@
+            EllipsoidRadiusSemiMajor = ellipseDetectedInPlane.Min(x => x.rMajor);
+            EllipsoidRadiusMajor = ellipseDetectedInPlane.Max(x => x.rMajor);
+        }
+        else if (rectangleDetectedInPlane.Count(x => x.y) == 3) // Cuboid detected
+        {
+            DetectedGeometry = PrimitiveGeometry.Cuboid;
+            float shortestVerticalEdge = rectangleDetectedInPlane.Min(x => x.vMax - x.vMin);
+            float shortestHorizontalEdge = rectangleDetectedInPlane.Min(x => x.hMax - x.hMin);
+            float longestVerticalEdge = rectangleDetectedInPlane.Max(x => x.vMax - x.vMin);
+            float longestHorizontalEdge = rectangleDetectedInPlane.Max(x => x.hMax - x.hMin);
+            CuboidShortestEdgeLength = Math.Min(shortestVerticalEdge, shortestHorizontalEdge);
+            CuboidIntermediateEdgeLength = Math.Max(shortestVerticalEdge, shortestHorizontalEdge);
+            CuboidLongestEdgeLength = Math.Max(longestVerticalEdge, longestHorizontalEdge);
+        }
+        else if (ellipseDetectedInPlane.Count(x => x.y) == 1 && rectangleDetectedInPlane.Count(x => x.y) == 2) // Cylinder detected
+        {
+            DetectedGeometry = PrimitiveGeometry.Cylinder;
+
+            var ellipsePlane = ellipseDetectedInPlane.Select((v, i) => new { v = v, i = i }).First(x => x.v.y);
+            CylinderRadiusMinor = ellipsePlane.v.rMinor;
+            CylinderRadiusMajor = ellipsePlane.v.rMajor;
+
+            var rectangleContainingHeight = rectangleDetectedInPlane[(ellipsePlane.i + 1) % 3];
+            float heightH = rectangleContainingHeight.hMax - rectangleContainingHeight.hMin;
+            float heightV = rectangleContainingHeight.vMax - rectangleContainingHeight.vMin;
+            CylinderHeight = Math.Max(heightH, heightV);
+        }
+        else if (!recursiveCall && ArePrincipalComponentsSymmetricInExactlyTwoAxis(pca)) // May be cuboid with one quadratic face
+        {
+            (int equal1, int equal2, int different) = DetermineEqualAndDifferentPrincipalComponents(pca);
+            pca = RedoPcaAssumingCuboidWithOneQuadraticFace(pca, vertices, different, equal1, equal2);
+            DoPrimitiveGeometryDetection(vertices, pca, true);
+        }
+        else if (!recursiveCall && ArePrincipalComponentsSymmetricInAllAxis(pca)) // May be a cube (i.e., all quadratic faces)
+        {
+            pca = RedoPcaAssumingACube(pca, vertices);
+            DoPrimitiveGeometryDetection(vertices, pca, true);
+        }
+        else
+        {
+            DetectedGeometry = PrimitiveGeometry.Unknown;
+            CylinderRadiusMinor = 0.0f;
+            CylinderRadiusMajor = 0.0f;
+            CylinderHeight = 0.0f;
+        }
+    }
+
+    private static (bool y, float hMin, float hMax, float vMin, float vMax) IsRectangle(
+        List<Vector3> local,
+        PcaResult3 pca,
+        List<(float a, float b)> projections,
+        int ua,
+        int ub
+    )
```

Comment:
Nit, but instead of passing Lists I would take Arrays or IImmutableList<Vector3> as inputs so its easier to see that no mutation happens.